hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/RequestsHandler.java

```diff
@@ -0,0 +1,558 @@
+ * configuration and can be updated dynamically without restarting the RM.
+ */
+public class RequestsHandler {
+
+  protected static final Logger LOG =
+      LoggerFactory.getLogger(RequestsHandler.class);
+
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  private static final ScriptEngine SCRIPT_ENGINE =
+      new ScriptEngineManager().getEngineByName("JavaScript");
+
+  private final RMContext rmContext;
+  private final ReentrantReadWriteLock.WriteLock writeLock;
+  private final ReentrantReadWriteLock.ReadLock readLock;
+
+  private boolean enabled = false;
+
+  private List<UpdateItem> updateItems;
+
+  // current updates conf value for comparing
+  private String updatesConfV;
+
+  public RequestsHandler(RMContext rmContext) {
+    this.rmContext = rmContext;
+    ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    writeLock = lock.writeLock();
+    readLock = lock.readLock();
+  }
+
+  public void initialize(Configuration conf)
+      throws IOException, YarnException {
+    if (SCRIPT_ENGINE == null) {
+      // disabled if script engine is not found
+      LOG.warn("Disabled RequestsHandler since script engine not found");
+      return;
+    }
+    boolean newEnabled =
+        conf.getBoolean(CapacitySchedulerConfiguration.REQUEST_HANDLER_ENABLED,
+            CapacitySchedulerConfiguration.DEFAULT_REQUEST_HANDLER_ENABLED);
+    List<UpdateItem> newUpdateItems = null;
+    String newUpdatesConfV = null;
+    if (newEnabled) {
+      newUpdatesConfV =
+          conf.get(CapacitySchedulerConfiguration.REQUEST_HANDLER_UPDATES);
+      UpdatesConf newUpdatesConf = null;
+      if (newUpdatesConfV != null && !newUpdatesConfV.isEmpty()) {
+        newUpdatesConf =
+            OBJECT_MAPPER.readValue(newUpdatesConfV, UpdatesConf.class);
+      }
+      if (newUpdatesConf != null && newUpdatesConf.getItems() != null &&
+          !newUpdatesConf.getItems().isEmpty()) {
+        newUpdateItems = new ArrayList<>();
+        for (UpdateItemConf updateItemConf : newUpdatesConf.getItems()) {
+          newUpdateItems.add(new UpdateItem(updateItemConf));
+        }
+      }
+    }
+    // update
+    writeLock.lock();
+    try{
+      if (enabled == newEnabled &&
+          StringUtils.equals(newUpdatesConfV, updatesConfV)) {
+        LOG.info("No changes detected in RequestsHandler configuration," +
+            " enabled={}, updatesConf={}", enabled, updatesConfV);
+        return;
+      }
+      enabled = newEnabled;
+      updateItems = newUpdateItems;
+      updatesConfV = newUpdatesConfV;
+      LOG.info("Initialized request updater, enabled={}, updatesConf={}",
+          enabled, updatesConfV);
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  public RequestsHandleResponse handle(FiCaSchedulerApp app,
+      List<ResourceRequest> resourceRequests,
+      List<SchedulingRequest> schedulingRequests) {
+    readLock.lock();
```

Comment:
This is an update operator, why use readLock?

Reply:
The readLock and writeLock are used to protect fields in RequestsHandler such as enabled and updateItems, which can be used concurrently, for example, they could be updated while refreshQueues is called in a RPC thread, meanwhile could be read to handle requests in other RPC threads. So that these fields should be modified exclusively in the initialize method (requiring writeLock for mutual exclusion) but only read concurrently in the handle method (where readLock enables multiple threads to access the shared state simultaneously without blocking)‌.